added securesettings handling

pschulten opened the pull request:
Add support for secure_settings in alertchanel resource implementation

Secure Settings:
https://grafana.com/docs/grafana/latest/administration/provisioning/#supported-settings 

https://registry.terraform.io/providers/grafana/grafana/latest/docs/resources/alert_notification#example-usage

```
settings = {​​​​​​​
  uploadImage = "false"
  apiUrl = "https://api.opsgenie.com/v2/alerts"
}​​​​​​​
secure_settings = {​​​​​​​
  apiKey = each.value
}​​​​​​​
```

Avoid having apiKeys and passwords exposed via UI if using tf provider to configure grafana alertnotifications

needs: https://github.com/grafana/grafana-api-golang-client/pull/21

trotttrotttrott reviewed (commented):
Thanks @pschulten. Your [API PR](https://github.com/grafana/grafana-api-golang-client/pull/21) was merged. Updating vendoring should make this pass.

```
go get github.com/grafana/grafana-api-golang-client@master
go mod tidy
go mod vendor
```

Could you also add this argument to the docs? https://github.com/grafana/terraform-provider-grafana/blob/4edbb7527c2f9b1c7c982f5e6681fbf2bc16334d/website/docs/r/alert_notification.html.md

Lastly, why do you need to convert boolean values to strings?

nouseforaname commented:
Hey,
Updated commits for docs and vendor should come in soon.. 

I'm not quite sure what you mean by `Lastly, why do you need to convert boolean values to strings?` We pretty much just copied over the code from settings handling to keep the code similar.. 

https://github.com/grafana/terraform-provider-grafana/blob/9494226cff02185e5a4cc5c17e3fc36cfabc8ccd/grafana/resource_alert_notification.go#L129-L139
vs
https://github.com/grafana/terraform-provider-grafana/blob/9494226cff02185e5a4cc5c17e3fc36cfabc8ccd/grafana/resource_alert_notification.go#L140-L151

https://github.com/grafana/terraform-provider-grafana/blob/9494226cff02185e5a4cc5c17e3fc36cfabc8ccd/grafana/resource_alert_notification.go#L186-L196
vs.
https://github.com/grafana/terraform-provider-grafana/blob/9494226cff02185e5a4cc5c17e3fc36cfabc8ccd/grafana/resource_alert_notification.go#L197-L207

What exactly would you like to see changed?

trotttrotttrott commented:
@nouseforaname ah okay, I didn't notice that was already how `settings` is implemented. This explains it: https://github.com/grafana/terraform-provider-grafana/pull/37. Let's leave it how you've got it. I'll follow up and see if we can eliminate the extra string/boolean conversion logic.

trotttrotttrott commented on vendor/modules.txt:
```diff
@@ -74,7 +74,11 @@ github.com/google/go-cmp/cmp/internal/function
+# github.com/grafana/grafana-api-golang-client v0.0.0-20210218192924-9ccd2365d2a6
+=======
 # github.com/grafana/grafana-api-golang-client v0.0.0-20210204123314-0aa501c4b8c4
+>>>>>>> github/master
```
```suggestion
# github.com/grafana/grafana-api-golang-client v0.0.0-20210218192924-9ccd2365d2a6
```

trotttrotttrott reviewed (changes requested):
Getting there! We're missing the `go.mod` and `go.sum` changes. Could you run the following again and commit them?

```
go get github.com/grafana/grafana-api-golang-client@master
go mod tidy
```

pschulten commented:
@trotttrotttrott sorry for the hassle :(

tenstad commented:
Anything blocking this one, or could it be merged if the conflicts were resolved?  
I really want this feature! Please tell me if there is anything I could do to speed up the process.

torees commented:
Anything I can do to help this to merge? I would like to get this feature into the provider

pschulten committed:
Add support for secure_settings in alertchanel resource implementation

Secure Settings:
https://grafana.com/docs/grafana/latest/administration/provisioning/#supported-settings

https://registry.terraform.io/providers/grafana/grafana/latest/docs/resources/alert_notification#example-usage

```
settings = {
  uploadImage = "false"
  apiUrl = "https://api.opsgenie.com/v2/alerts"
}
secure_settings = {
  apiKey = each.value
}
```

Avoid having apiKeys and passwords exposed via UI if using tf provider to configure grafana alertnotifications

needs: https://github.com/grafana/grafana-api-golang-client/pull/21

pschulten committed:
Add support for secure_settings in alertchanel resource implementation

Secure Settings:
https://grafana.com/docs/grafana/latest/administration/provisioning/#supported-settings

https://registry.terraform.io/providers/grafana/grafana/latest/docs/resources/alert_notification#example-usage

```
settings = {
  uploadImage = "false"
  apiUrl = "https://api.opsgenie.com/v2/alerts"
}
secure_settings = {
  apiKey = each.value
}
```

Avoid having apiKeys and passwords exposed via UI if using tf provider to configure grafana alertnotifications

needs: https://github.com/grafana/grafana-api-golang-client/pull/21

tenstad commented on grafana/resource_alert_notification.go:
```diff
@@ -57,6 +57,11 @@ func ResourceAlertNotification() *schema.Resource {
 				Optional:  true,
 				Sensitive: true,
```
With the addition of `secure_settings`, could `settings` be `Sensitive: false`?

nouseforaname replied:
People might still decide to put the apikey in settings

tenstad replied:
It would be nice to see the diffs in `settings` in the plan, but I see your point.

thor replied:
People *might* still decide to commit API keys as well. 😋 

People *should* not place API keys in `settings` once `secure_settings` is available, right? 
If `settings.Sensitive` remains `true`, it's quite hard to see any non-sensitive changes, such as adding or removing e-mail addresses.

If you use Terraform 0.14 with the enabled experiment, or you've upgraded to 0.15 and [sensitive provider variables are hidden by default (and cannot be overridden)](https://www.terraform.io/docs/extend/best-practices/sensitive-state.html#using-the-sensitive-flag), you *could* reasonably expect that your secret values remain hidden in the CLI output. Fair enough. 

**Alternative**: bump the major version.

You could consider the possibility of exposing sensitive values in the CLI a breaking change, even though 0.15 hasn't been out for long, and allow those who expect non-encrypted, non-sensitive and sensitive values alike to remain hidden to continue to do so. 
If they choose to upgrade, they would (hopefully) read the release notes informing them that they must move their sensitive values to `secure_settings` if they wish for them to remain hidden.
That way, those who want to see their non-sensitive changes can do so, and we can have our cake and eat it too.

Then again, people *might* still decide to do a major version upgrade without reading the release notes.

nouseforaname replied:
Hmm. I'd opt for a warning first (like you see with the azure provider) and then mark it false with some time to adjust. Let's be hones here, 90% of people do not read the release notes.

nouseforaname replied:
@trotttrotttrott  any preferences on the above? I can add it to the PR if you want.

nouseforaname commented:
@trotttrotttrott We're kinda running in circles.. Can the conflicts just be resolved on merge? We update the PR, the main branch gets updated, we're updating the PR again :)

nouseforaname commented:
Hey,

just wanted to ask if and how we want to proceed with this? Do you required the changes discussed above?

trotttrotttrott commented:
Sorry for the delay on this. I just moved this to https://github.com/grafana/terraform-provider-grafana/pull/208 so I could resolve the conflicts and refactor the documentation style.

tenstad commented:
Thank you 👍  What is your opinion regarding `settings` being non-sensitive now that `secure_settings` are added? See the discussion earlier in the MR.
